src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs

```diff
@@ -255,26 +245,22 @@ internal partial class CommonReferenceManager<TCompilation, TAssemblySymbol>
                             continue;
                         }
 
-                        if (!requestedIdentities.Add(binding.ReferenceIdentity))
+                        if (!TryResolveMissingReference(
+                            requestingReference,
+                            binding.ReferenceIdentity,
+                            ref implicitReferenceResolutions,
+                            resolver,
+                            resolutionDiagnostics,
+                            out AssemblyIdentity resolvedAssemblyIdentity,
+                            out AssemblyMetadata resolvedAssemblyMetadata,
+                            out PortableExecutableReference resolvedReference))
                         {
+                            resolutionFailures.Add(binding.ReferenceIdentity);
                             continue;
                         }
 
-                        PortableExecutableReference resolvedReference;
-                        if (previouslyResolvedAssembliesOpt == null || !previouslyResolvedAssembliesOpt.TryGetValue(binding.ReferenceIdentity, out resolvedReference))
-                        {
-                            resolvedReference = resolver.ResolveMissingAssembly(requestingReference, binding.ReferenceIdentity);
-                            if (resolvedReference == null)
-                            {
-                                continue;
-                            }
-                        }
-
-                        var data = ResolveMissingAssembly(binding.ReferenceIdentity, resolvedReference, importOptions, resolutionDiagnostics);
-                        if (data == null)
-                        {
-                            continue;
-                        }
+                        // one attempt for resolution succeeded (the attempt is cached in implicitReferenceResolutions, so further attempts won't fail and add it back):
+                        resolutionFailures.Remove(binding.ReferenceIdentity);
```

Comment:
>resolutionFailures.Remove(binding.ReferenceIdentity); [](start = 24, length = 53)

It would be good to add an explanation how could we first fail and then succeed for the same identity. We don't need to go back and fix something up for bindings it failed?

Reply:
OK. The resolver takes 2 parameters: the metadata reference that references the identity and the identity. The default implementation of the resolver looks up the missing reference in the directory containing the metadata reference. Suppose the same identity is referenced from two metadata references from different directories and only one directory contains the assembly of the requested identity. Then one resolution will fail and the other will succeed. The failed resolutions are not "committed" until the end of the resolution, so we don't need to fix up anything.

Reply:
The comment on TryResolveMissingReference explains. Maybe I should move it to the caller?

Reply:
Added clarification.